File: app/schemas/gft.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class GFTPrincipioActivoRef(BaseModel):
    id: UUID | None = None
    slug: str | None = None
    nombre: str


class GFTPrincipioActivoIndexItem(BaseModel):
    id: UUID
    slug: str
    nombre: str
    letra: str
    count: int


class GFTPrincipioActivoIndexResponse(BaseModel):
    items: list[GFTPrincipioActivoIndexItem] = Field(default_factory=list)


class GFTAtcRef(BaseModel):
    codigo: str
    nombre: str | None = None
    nivel: str | None = None


class GFTAtcIndexItem(BaseModel):
    codigo: str
    nombre: str | None = None
    nivel: str
    count: int


class GFTAtcIndexResponse(BaseModel):
    items: list[GFTAtcIndexItem] = Field(default_factory=list)


class GFTDocumentoCimaRef(BaseModel):
    tipo: int | str | None = None
    url: str | None = None
    urlHtml: str | None = None
    secc: str | None = None
    fecha: str | None = None
    titulo: str | None = None
    nombre: str | None = None


class GFTIndicacionAutorizadaBifimed(BaseModel):
    indicacion_autorizada: str | None = None
    situacion_expediente_indicacion: str | None = None
    resolucion_expediente_financiacion_indicacion: str | None = None
    financiada: bool | None = None


class GFTFinanciacionDetalle(BaseModel):
    situacion_financiacion: str | None = None
    condiciones_financiacion_restringidas: str | None = None
    condiciones_especiales_financiacion: str | None = None
    estado_nomenclator: str | None = None
    aportacion_usuario: str | None = None
    subgrupo_atc: str | None = None
    last_synced_at: datetime | None = None
    indicaciones_autorizadas: list[GFTIndicacionAutorizadaBifimed] | None = None


class GFTClinicalSummaryAuto(BaseModel):
    source_status: str | None = None
    generated_at: datetime | None = None
    resumen_general: str | None = None
    indicaciones: str | None = None
    posologia: str | None = None
    ajuste_renal: str | None = None
    ajuste_hepatico: str | None = None
    contraindicaciones: str | None = None
    advertencias: str | None = None
    embarazo: str | None = None
    lactancia: str | None = None
    fuentes: dict | list | None = None
    warnings: list | dict | None = None


class GFTMedicamentoListItem(BaseModel):
    cn: str
    nombre: str | None = None
    presentacion: str | None = None
    forma_farmaceutica: str | None = None
    forma_farmaceutica_simplificada: str | None = None
    vias_administracion: list[str] = Field(default_factory=list)
    atc: list[GFTAtcRef] = Field(default_factory=list)
    principios_activos: list[GFTPrincipioActivoRef] = Field(default_factory=list)
    nemonico: str | None = None
    restricciones_hospitalarias: str | None = None
    ajuste_insuficiencia_renal: str | None = None
    ajuste_insuficiencia_hepatica: str | None = None
    precauciones_embarazo: str | None = None
    precauciones_lactancia: str | None = None
    situacion_financiacion: str | None = None
    url_ficha_tecnica: str | None = None
    url_prospecto: str | None = None
    fecha_ficha_tecnica: date | None = None
    fecha_prospecto: date | None = None
    indicaciones_ficha_tecnica: str | None = None


class GFTMedicamentoDetail(GFTMedicamentoListItem):
    observaciones_publicables: str | None = None
    documentos: list[GFTDocumentoCimaRef] = Field(default_factory=list)
    financiacion_detalle: GFTFinanciacionDetalle | None = None
    resumen_clinico_auto: GFTClinicalSummaryAuto | None = None


class GFTListResponse(BaseModel):
    total: int
    limit: int
    offset: int
    items: list[GFTMedicamentoListItem] = Field(default_factory=list)

File: frontend/src/types/gft.ts
export interface GFTPrincipioActivoRef {
  id: string | null;
  slug: string | null;
  nombre: string;
}

export interface GFTAtcRef {
  codigo: string;
  nombre: string | null;
  nivel: string | null;
}


export interface GFTCanonicalIndicacionBifimed {
  indicacion_autorizada: string | null;
  situacion_expediente_indicacion: string | null;
  resolucion_expediente_financiacion_indicacion: string | null;
  financiada?: boolean | null;
}

export interface GFTCanonicalPayload {
  cn: string;
  nombre_comercial: string | null;
  incluido_gft?: boolean | null;
  publicado?: boolean | null;
  observaciones_revision?: string | null;
  estado_publicacion?: string | null;
  principio_activo: string | null;
  forma_farmaceutica: string | null;
  via_administracion: string | null;
  nemonico: string | null;
  codigo_atc: string | null;
  descripcion_atc: string | null;
  jerarquia_atc: GFTAtcRef[];
  indicaciones_ficha_tecnica: string | null;
  url_ficha_tecnica?: string | null;
  url_prospecto?: string | null;
  estado_cima: 'disponible' | 'no_informado' | string;
  bifimed_cache_presente: boolean;
  situacion_financiacion_bifimed: string | null;
  condiciones_financiacion_restringidas: string | null;
  condiciones_especiales_financiacion: string | null;
  detalle_financiacion_json?: unknown;
  indicaciones_bifimed: GFTCanonicalIndicacionBifimed[];
  estado_bifimed: 'disponible' | 'no_informado' | 'sin_cache' | 'sin_indicaciones' | string;
  ajuste_insuficiencia_renal: string | null;
  ajuste_insuficiencia_hepatica: string | null;
  precauciones_embarazo: string | null;
  precauciones_lactancia: string | null;
  restricciones_hospitalarias: string | null;
  resumen_clinico_auto?: GFTResumenClinicoAuto | null;
  estado_resumen_clinico?: string | null;
  fuentes_disponibles: string[];
  campos_faltantes: string[];
  warnings: string[];
  data_quality_flags: string[];
}

export interface GFTMedicamentoListItem {
  cn: string;
  nombre: string | null;
  presentacion: string | null;
  forma_farmaceutica: string | null;
  forma_farmaceutica_simplificada: string | null;
  vias_administracion: string[];
  atc: GFTAtcRef[];
  principios_activos: GFTPrincipioActivoRef[];
  nemonico: string | null;
  indicaciones_ficha_tecnica: string | null;
  restricciones_hospitalarias: string | null;
  ajuste_insuficiencia_renal: string | null;
  ajuste_insuficiencia_hepatica: string | null;
  precauciones_embarazo: string | null;
  precauciones_lactancia: string | null;
  situacion_financiacion: string | null;
  url_ficha_tecnica: string | null;
  url_prospecto: string | null;
  fecha_ficha_tecnica: string | null;
  fecha_prospecto: string | null;
}

export interface GFTDocumentoCimaRef {
  tipo: number | string | null;
  url: string | null;
  urlHtml: string | null;
  secc: string | null;
  fecha: string | null;
  titulo: string | null;
  nombre: string | null;
}

export interface GFTIndicacionAutorizadaBifimed {
  indicacion_autorizada: string | null;
  situacion_expediente_indicacion: string | null;
  resolucion_expediente_financiacion_indicacion: string | null;
  financiada: boolean | null;
}

export interface GFTFinanciacionDetalle {
  situacion_financiacion: string | null;
  condiciones_financiacion_restringidas: string | null;
  condiciones_especiales_financiacion: string | null;
  estado_nomenclator: string | null;
  aportacion_usuario: string | null;
  subgrupo_atc: string | null;
  last_synced_at?: string | null;
  indicaciones_autorizadas?: GFTIndicacionAutorizadaBifimed[] | null;
}
export interface GFTResumenClinicoAuto {
  source_status: string;
  generated_at: string | null;
  resumen_general: string | null;
  indicaciones: string | null;
  posologia: string | null;
  ajuste_renal: string | null;
  ajuste_hepatico: string | null;
  contraindicaciones: string | null;
  advertencias: string | null;
  embarazo: string | null;
  lactancia: string | null;
  fuentes: Record<string, string[]>;
  warnings: string[];
}

export interface GFTMedicamentoDetail extends GFTMedicamentoListItem {
  observaciones_publicables: string | null;
  documentos: GFTDocumentoCimaRef[];
  financiacion_detalle: GFTFinanciacionDetalle | null;
  resumen_clinico_auto?: GFTResumenClinicoAuto | null;
  canonical_payload?: GFTCanonicalPayload | null;
}

export interface GFTListResponse {
  total: number;
  limit: number;
  offset: number;
  items: GFTMedicamentoListItem[];
}

export interface GFTAtcIndexItem {
  codigo: string;
  nombre: string | null;
  nivel: string;
  count: number;
}

export interface GFTAtcIndexResponse {
  items: GFTAtcIndexItem[];
}

export interface GFTPrincipioActivoIndexItem {
  id: string;
  slug: string;
  nombre: string;
  letra: string;
  count: number;
}

export interface GFTPrincipioActivoIndexResponse {
  items: GFTPrincipioActivoIndexItem[];
}

export interface ListMedicamentosParams {
  q?: string;
  letra?: string;
  principio_activo?: string;
  atc?: string;
  limit: number;
  offset: number;
}
